int16 state
int8 connected